### PR TITLE
Extended metrics

### LINKED ISSUE #454: Use testing short flag to skip playback tests
Using `go test -short` will skip the gateway playback testing.
It will enable faster basic test run without download and play of playback data.

### LINKED ISSUE #4265: Extended analytics event
Extended analytics enables reporting more information on each event report.
Only when relevant the following values, hashed except 'client', will be reported:

1. Repository - hashed value of the repository name
2. Ref - hashed value of the tag/commit/branch name 
3. SourceRef - hashed value of source tag/commit/branch 
4. UserID - hashed value of the user ID
5. Client - user agent of the client performing the operation

This feature is turned **off** by default.

### PR DESCRIPTION
**Pull request by nopcoder**
- default is off.
- sends client user-agent as id, repository, ref, source ref and username.
- all extended values are hashed, except client id.
- lakectl, webui and sdk now include specific user-agent with build time version information.

Close #4265

### CODE REVIEW

File: clients/java/src/main/java/io/lakefs/clients/api/ApiClient.java
Comment (itaiad200) [inline clients/java/src/main/java/io/lakefs/clients/api/ApiClient.java:134]: Should we check this in? IIUC this will now change with every build.

File: clients/python/lakefs_client/api_client.py
Comment (itaiad200) [inline clients/python/lakefs_client/api_client.py:80]: Same here

File: webui/vite.config.js
Comment (itaiad200) [inline webui/vite.config.js:9]: Sorcery

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Partial review, will continue tomorrow.

File: clients/java/src/main/java/io/lakefs/clients/api/ApiClient.java
Comment (nopcoder) [reply clients/java/src/main/java/io/lakefs/clients/api/ApiClient.java:134]: You are right I'll keep just the VERSION value - we will have 'dev' and every publish we will have the lakefs version without the git sha.

File: webui/vite.config.js
Comment (nopcoder) [reply webui/vite.config.js:9]: Had to find a quick way to embed this information inside our app.

File: pkg/api/controller.go
Comment (NoyDavidson2) [reply pkg/api/controller.go:2199]: AFAIC it's not at the moment. 
Might be in the future - if we collect commit id in other actions like create commit and not just at the get commit point.

Comment (nopcoder) [reply pkg/api/controller.go:2199]: @itaiad200 it is also two for other operations - diff, merge and create branch.

File: cmd/lakefs/cmd/run.go
Comment (nopcoder) [reply cmd/lakefs/cmd/run.go:107]: Will drop the code - it helpful in the development time of this feature.

File: pkg/api/controller.go
Comment (NoyDavidson2) [reply pkg/api/controller.go:2199]: @nopcoder create branch only gets 'branch' as ref according to the table you added on the issue associated #454

@itaiad200 I agree about the commit ID at this point, it's not part of the requirements, let's consider not using it in this constellation.
I share your concern about multiple values with different meaning for this field (ref/branch/commit ID).
Going back to the original requirement - we wanted to collect branch, and then came up the question - what branch do we keep for events like 'merge'? we do need both source and destination. (similar case with diff)
This is why came up the idea of additional field 'source ref', but now I'm not so sure keeping multiple fields for these unique cases seems reasonable.

I don't want to block this PR, I think we should stick to the original plan and add branch field with branch values **only** at this point, going forward if we see fit we can add the others in some form.

As for the 'merge' events I'm thinking maybe we can keep the **source branch** as 'branch' value and create another table for merges with a foreign key field, on the 'events' table side, referring to 'merge'(installation id, time, source, destination) table. On the 'merge' table we can keep any other information about the merge if there is. 
Makes sense? WDYT?

Comment (itaiad200) [reply pkg/api/controller.go:2199]: I see many BI collections that use the `ref` as is. It's more than the `CommitID` being probably redundant and will hard to deduct any meaningful insights from it. It also masks important branch information. However I understand the limitations - `Ref` interpolation is happening in Catalog/Graveler, and not in the Controller where BI is collected. I'm fine with collecting it as it's in this PR, as long as @NoyDavidson2 is onboard with collecting `Refs` which is a fancy word for either Branch/Commit/Traverse of the commit tree (`main^2~3`)

File: pkg/api/controller.go
Comment (nopcoder) [reply pkg/api/controller.go:2734]: We can update this one if we agree to also run 'update' on the current database after we deploy

File: pkg/gateway/operations/listbuckets.go
Comment (nopcoder) [reply pkg/gateway/operations/listbuckets.go:24]: For each handler 'o' got different type - there are at least 3.

File: pkg/api/controller.go
Comment (nopcoder) [reply pkg/api/controller.go:2205]: 1. Got the linked issue wrong. The number is from a different repo - will probably need to replace it with non link or a different issue.
2. There is no point in linking another table with extra properties for each event, unless we create different tables for different types as the relation is 1:1 + it will require adding unique ID to the event to bind the extra information.
3. Wanted to make it clear, as @itaiad200 said - almost all places that we accept 'ref' and not 'branch' can report commit id (event short), branch, tag or relative commit which will make the different IDs to work with.
If we plan to use just branch names - at the point we report the metric we can limit the branch name just to operations that accepts branch name. It in case of 'merge' the target branch name. But in case of 'diff' we don't have any promise we compare two branches.

File: pkg/stats/collector.go
Comment (itaiad200) [inline pkg/stats/collector.go:379]: 2 hex digits gives you 256 options which is considered safe usage for 16 unique examples. I don't think it's enough for branches, let alone commits which can accumulate quite heavily.

**Review by itaiad200 — APPROVED** [review]
See single comment before merging

File: pkg/stats/collector.go
Comment (nopcoder) [reply pkg/stats/collector.go:379]: This is full hex string with leading zero for 2 digit padding it doesn't cut the value

File: pkg/api/controller.go
Comment (NoyDavidson2) [reply pkg/api/controller.go:2205]: Update - @nopcoder and I have spoke this morning and decided to release this feature the way it is described here. After working with the data we will decide if and how we can make it better as we go.

File: webui/src/lib/api/index.js
Comment (itaiad200) [inline webui/src/lib/api/index.js:143]: Is `JSON.stringify(data, null, "")` equal to `JSON.stringify(data)` ?

File: pkg/stats/collector.go
Comment (itaiad200) [inline pkg/stats/collector.go:247]: Why would run be called more than once?

Comment (itaiad200) [reply pkg/stats/collector.go:247]: Also consider using `Sync.Once`.

File: pkg/httputil/client.go
Comment (itaiad200) [inline pkg/httputil/client.go:10]: Won't it spam the DB? The cardinality of user-agent is huge

**Review by itaiad200 — APPROVED** [review]

File: webui/src/lib/api/index.js
Comment (nopcoder) [reply webui/src/lib/api/index.js:143]: Yea, same.

File: pkg/stats/collector.go
Comment (nopcoder) [reply pkg/stats/collector.go:247]: Will do. I also think it is bit over-eng - but will address it in a different issue.

Comment (nopcoder) [reply pkg/stats/collector.go:247]: sync.Once can help making sure `Run` will perform per incese. Currently the counter also used a marker for 'Close' to indicate we do not close/wait for resources we allocated on Run.

File: pkg/httputil/client.go
Comment (nopcoder) [reply pkg/httputil/client.go:10]: I consider 'event' information as raw data - we can filter out the data. Using the user-agent as the client identify will help us also process previous/older clients using the lakefs.

**Pull request merged**
